### PR TITLE
Document array length changing behaviors

### LINKED ISSUE #4802: Does anArray.length-- delete the last element?
Normally I send questions to StackOverflow.

But here I see an assertion that calling `.length--` on an array will delete the last element and recover gas.

> https://github.com/OpenZeppelin/openzeppelin-solidity/blob/master/contracts/token/ERC721/ERC721Token.sol#L151

The people on that team are generally smart so I'm going to assume it is true.

If it is true then can the documentation please be update to explain this feature?

### PR DESCRIPTION
**Pull request by fulldecent**
Fixes #4802 

### Your checklist for this pull request

Please review the [guidelines for contributing](http://solidity.readthedocs.io/en/latest/contributing.html) to this repository.

Please also note that this project is released with a [Contributor Code of Conduct](CONDUCT.md). By participating in this project you agree to abide by its terms.

### Checklist
- [ ] Code compiles correctly
- [ ] All tests are passing
- [ ] New tests have been created which fail without the change (if possible)
- [ ] README / documentation was extended, if necessary
- [ ] Changelog entry (if change is visible to the user)
- [x] Used meaningful commit messages

### Description
Please explain the changes you made here.

Thank you for your help!

### CODE REVIEW

File: docs/types.rst [hunk 1]
```diff
@@ -751,13 +751,12 @@ Members
 ^^^^^^^
 
 **length**:
-    Arrays have a ``length`` member to hold their number of elements.
-    Dynamic arrays can be resized in storage (not in memory) by changing the
-    ``.length`` member. This does not happen automatically when attempting to access elements outside the current length. The size of memory arrays is fixed (but dynamic, i.e. it can depend on runtime parameters) once they are created.
+    Fixed-length arrays (storage or memory) have a read-only ``length`` member to access their number of elements.
+    Dynamic-sided arrays (only available for storage) have a read-write ``length`` member to resize the arary. Increasing the length adds unitialized elements to the array, this is O(1) complexity. Reducing the length performs ``delete()`` on each removed element and is O(n) complexity proportional to the number of elements being deleted. Please note that calling ``length--`` on an empty array will set the length of the array to 2^256-1 due to uint256 underflow wrapping.
```
Comment (leonardoalt) [inline docs/types.rst]: The spellchecker caught `unitialized  ==> uninitialized` here.

Comment (chriseth) [inline docs/types.rst]: Typo: "sided"

Comment (chriseth) [inline docs/types.rst]: Perhaps rather `O(n) proportional to...` use `O(n) where 'n' is the number of elements deleted`

Comment (chriseth) [inline docs/types.rst]: Actually `delete` is not a function (also below).

Comment (leonardoalt) [reply docs/types.rst]: There are other typos, such as:
`dynamic-sided => dynamically-sized`
`arary => array`

Comment (leonardoalt) [reply docs/types.rst]: You can also have `memory` dynamically-sized arrays, just not resize them that way.

Comment (fulldecent) [reply docs/types.rst]: Fixed typos here in a commit with a message that appropriately, but unintentionally, includes the same 2b1502f22bfb992ff61de12619f0d7c2526ff5f1

Comment (fulldecent) [reply docs/types.rst]: Fixed typos here in a commit with a message that appropriately, but unintentionally, includes the same 2b1502f22bfb992ff61de12619f0d7c2526ff5f1

Comment (fulldecent) [reply docs/types.rst]: Updated in 2b1502f22bfb992ff61de12619f0d7c2526ff5f1

Comment (fulldecent) [reply docs/types.rst]: Updated here 2b1502f22bfb992ff61de12619f0d7c2526ff5f1

Comment (fulldecent) [reply docs/types.rst]: Are you talking about memory arrays that have a length specified in runtime?

Comment (chriseth) [reply docs/types.rst]: Yes, we usually call them dynamically-sized arrays, even if you cannot change their length once created.

File: docs/types.rst [hunk 2]
```diff
@@ -752,11 +752,11 @@ Members
 
 **length**:
     Fixed-length arrays (storage or memory) have a read-only ``length`` member to access their number of elements.
-    Dynamic-sided arrays (only available for storage) have a read-write ``length`` member to resize the arary. Increasing the length adds unitialized elements to the array, this is O(1) complexity. Reducing the length performs ``delete()`` on each removed element and is O(n) complexity proportional to the number of elements being deleted. Please note that calling ``length--`` on an empty array will set the length of the array to 2^256-1 due to uint256 underflow wrapping.
+    Dynamically-sized arrays (only available for storage) have a read-write ``length`` member to resize the array. Increasing the length adds uninitialized elements to the array, this is O(1) complexity. Reducing the length performs ``delete`` on each removed element and is O(n) complexity where n is the number of elements being deleted. Please note that calling ``length--`` on an empty array will set the length of the array to 2^256-1 due to uint256 underflow wrapping.
```
Comment (chriseth) [inline docs/types.rst]: Is it "is" or "has" (complexity)?

Comment (fulldecent) [reply docs/types.rst]: Fixed in 61286b341a63146b1dd943dae251c38da0280666

File: docs/types.rst [hunk 3]
```diff
@@ -751,13 +751,12 @@ Members
 ^^^^^^^
 
 **length**:
-    Arrays have a ``length`` member to hold their number of elements.
-    Dynamic arrays can be resized in storage (not in memory) by changing the
-    ``.length`` member. This does not happen automatically when attempting to access elements outside the current length. The size of memory arrays is fixed (but dynamic, i.e. it can depend on runtime parameters) once they are created.
+    Fixed-length arrays (storage or memory) have a read-only ``length`` member to access their number of elements.
+    Dynamically-sized arrays (only available for storage) have a read-write ``length`` member to resize the array. Increasing the length adds uninitialized elements to the array, this has O(1) complexity. Reducing the length performs ``delete`` on each removed element and is O(n) complexity where n is the number of elements being deleted. Please note that calling ``length--`` on an empty array will set the length of the array to 2^256-1 due to uint256 underflow wrapping.
```
Comment (chriseth) [inline docs/types.rst]: This could be read as if dynamically-sized arrays do not exist in memory.

Comment (fulldecent) [reply docs/types.rst]: It is correct that dynamically-sized arrays do not exist in memory.

Maybe we can use the language "variable length array" to describe run-time-configurable fixed-length in-memory arrays that Solidity uses. This is consistent with the language used in the C99 (ISO 9899:1999) standard.

Would this word choice be acceptable?

Comment (axic) [inline docs/types.rst]: Syntactical improvement:
- please put `O(1)`, `O(n)` and `n` into emphasis (eg. `*O(1)`)
- and `uint256` into backticks

File: docs/types.rst [hunk 4]
```diff
@@ -751,13 +751,12 @@ Members
...
 **push**:
      Dynamic storage arrays and ``bytes`` (not ``string``) have a member function called ``push`` that can be used to append an element at the end of the array. The function returns the new length.
 **pop**:
-     Dynamic storage arrays and ``bytes`` (not ``string``) have a member function called ``pop`` that can be used to remove an element from the end of the array.
+     Dynamic storage arrays and ``bytes`` (not ``string``) have a member function called ``pop`` that can be used to remove an element from the end of the array. This will also implicitly call ``delete`` on the removed element.
```
Comment (axic) [inline docs/types.rst]: Do we have a section for documenting delete properly? If so, can you link to it here and above?

Comment (fulldecent) [reply docs/types.rst]: Updated in 9169ac5

File: docs/types.rst (hunk 3 above)
Comment (fulldecent) [reply docs/types.rst]: Updated in 9169ac5

**Comment by axic** [comment]
This needs to be squashed.

File: docs/types.rst [hunk 5]
```diff
@@ -751,13 +751,12 @@ Members
 ^^^^^^^
 
 **length**:
-    Arrays have a ``length`` member to hold their number of elements.
-    Dynamic arrays can be resized in storage (not in memory) by changing the
-    ``.length`` member. This does not happen automatically when attempting to access elements outside the current length. The size of memory arrays is fixed (but dynamic, i.e. it can depend on runtime parameters) once they are created.
+    Arrays (storage or memory) have a ``length`` member to read their number of elements.
```
Comment (axic) [inline docs/types.rst]: I think the `storage or memory` addition is not useful here.

**Comment by chriseth** [comment]
Ah sorry, I also edited this part. If you don't mind, I will squash it and remove (storage or memory) as axic suggested.

**Comment by fulldecent** [comment]
Great, thank you.

**Pull request merged**